### **Exercice d'application**

Afin d'appliquer ce que vous venez d'apprendre, je vous propose de mener l'enquête avec Agatha Christie, en incarnant le Hercule Poirot du NLP pour démasquer le coupable...

<img src="https://www.citoyens-justice.fr/k-stock/data/visuel-site/investigation-judiciaire.png" alt="maj_to_min" width="400">

Pour cela, vous devrez commencer par télécharger [ce fichier](https://github.com/amandinejvl/ObTIC/blob/main/atelier_preprocessing/data/agatha_christie.txt) puis l'importer dans Google Drive.

Il s'agit du roman d'Agatha Christie _Ils étaient dix_ (1939).

#

<img src="https://static.wikia.nocookie.net/lunivers-dagatha-christie/images/1/13/Ils_%C3%A9taient_dix_%C3%A9dition_fran%C3%A7aise_.jpg/revision/latest?cb=20250314175546&path-prefix=fr" alt="maj_to_min" width="200">

**Mise en situation**

Vera Claythorne, institutrice, est engagée comme secrétaire auprès d'une certaine madame Owen, qui vit avec son époux sur l'île du Nègre, sur la côte du Devon, en Angleterre. En arrivant sur place, la jeune femme découvre qu'elle doit composer avec la présence de nombreux invités qui ne se connaissent pas. Les propriétaires, eux, sont mystérieusement absents. Au cours de la soirée, chaque convive est accusé d'un ou plusieurs meurtres.

Dans les quelques jours qui suivent, tous les invités, qui ne se connaissaient pas entre eux, meurent brutalement l'un après l'autre, à chaque fois d'une façon qui rappelle les couplets d'une comptine intitulée _Dix Petits Nègres_.

Afin d'élucider cette série de crimes, la police a décidé de faire appel à ses experts en NLP (vous !).

**💡 Indices :**

Les indices dont nous disposons sont les suivants :

- Le nom du coupable est présent dans au moins 50 phrases ;
- Hors ponctuation et mots vides, les phrases contenant le nom du coupable totalisent entre 500 et 1300 mots ;
- Hors ponctuation et mots vides, la partie du discours la plus fréquente dans les phrases contenant le nom du coupable n'est pas le verbe ;
- Hors ponctuation et mots vides, le lemme le plus fréquent dans les phrases contenant le nom du coupable ne commence pas par la lettre "_R_" ;
- Le verbe dont le suspect est le plus souvent le sujet ne commence pas par la lettre "_D_".


#

**_Suspects :_** _Armstrong, Blore, Brent, Claythorne, Lombard, Macarthur, Marston, Wargrave, Ethel Rogers, Thomas Rogers_

**Configuration**

In [2]:
# Config Colab
from google.colab import drive
drive.mount('/content/drive')

import sys
sys.path.append("/content/drive/MyDrive/Obtic/Ateliers/AT_spaCy") # A adapter selon votre chemin

Mounted at /content/drive


**Chargement du corpus**

In [3]:
# Chemin vers les données (adapter à vos données)
text_file = "/content/drive/MyDrive/Obtic/Ateliers/AT_spaCy/data/agatha_christie.txt"

import re

# Charger et afficher le corpus
with open(text_file, "r") as f:
    texte = re.sub(r"\n+", " ", f.read())
    texte = re.sub(r"\s+", " ", texte)
    texte = texte.strip()

texte

'Dans le coin fenêtre d’un compartiment fumeurs de première classe, le juge Wargrave, retraité depuis peu, tirait sur son cigare en parcourant avec intérêt les pages politiques du Times. Posant son journal, il regarda par la vitre. Ils traversaient maintenant le Somerset. Il jeta un coup d’œil à sa montre : encore deux heures de voyage. Mentalement, il passa en revue tout ce qui avait paru dans la presse au sujet de l’île du Nègre. Il y avait d’abord eu la nouvelle de son achat par un milliardaire américain fanatique de yachting – assortie de la description de la luxueuse demeure ultra-moderne qu’il faisait construire sur cet îlot au large du Devon. Le fait malencontreux que la toute récente et néanmoins troisième épouse dudit milliardaire n’eût pas le pied marin avait entraîné la mise en vente de l’île et de la maison. Des publicités dithyrambiques avaient alors été placardées un peu partout. Jusqu’au jour où on avait sobrement annoncé qu’elle avait été rachetée par un certain Mr O’Ny

**Traitement du texte avec spaCy**

In [5]:
# Installer spaCy
!pip install spacy

# Télécharger le modèle français fr_core_news_lg
!python3 -m spacy download fr_core_news_lg

# Importer spaCy
import spacy

# Charger le modèle
nlp = spacy.load('fr_core_news_lg')

# Traiter le texte avec spacy
doc = nlp(texte)

  Using cached https://github.com/explosion/spacy-models/releases/download/fr_core_news_lg-3.8.0/fr_core_news_lg-3.8.0-py3-none-any.whl (571.8 MB)
✔ Download and installation successful
You can now load the package via spacy.load('fr_core_news_lg')
⚠ Restart to reload dependencies
If you are in a Jupyter or Colab notebook, you may need to restart Python in
order to load all the package's dependencies. You can do this by selecting the
'Restart kernel' or 'Restart runtime' option.


💡 **Indice 1 :** Le nom du coupable est présent dans au moins 50 phrases.

**Fonctions utiles :**

- Extraction de phrases (`.sents`)
- Extraction de texte (`.text`)
- 2 boucles `for` imbriquées
- 1 condition d'appartenance (`if` ... `in` ...)

**Méthode :**

In [6]:
# Objectif : Compter le nombre de phrases dans lesquelles les noms des suspects apparaissent

# Noms des suspects
suspects = ["Armstrong", "Blore", "Brent", "Claythorne", "Lombard", "Macarthur", "Marston", "Wargrave", "Ethel Rogers", "Thomas Rogers"]

# Segmenter en phrases

# Parcourir chaque suspect
    # Initialiser le compteur de phrases dans lesquelles le nom du suspect apparaît
    # Parcourir chaque phrase
        # Si le nom du suspect est dans la phrase
            # Alors on ajoute 1 à son nombre de phrases

    # Affichage

**Solution :**

In [7]:
suspects = ["Armstrong", "Blore", "Brent", "Claythorne", "Lombard", "Macarthur", "Marston", "Wargrave", "Ethel Rogers", "Thomas Rogers"]

sentences = [sent.text for sent in doc.sents]

for suspect in suspects:
    nb_sentences_with_suspect = 0
    for sentence in sentences:
        if suspect in sentence:
            nb_sentences_with_suspect += 1

    print(f"Nombre de phrases contenant {suspect} : {nb_sentences_with_suspect}")

Nombre de phrases contenant Armstrong : 236
Nombre de phrases contenant Blore : 277
Nombre de phrases contenant Brent : 130
Nombre de phrases contenant Claythorne : 74
Nombre de phrases contenant Lombard : 300
Nombre de phrases contenant Macarthur : 56
Nombre de phrases contenant Marston : 55
Nombre de phrases contenant Wargrave : 113
Nombre de phrases contenant Ethel Rogers : 1
Nombre de phrases contenant Thomas Rogers : 1


➡️ On peut éliminer :
- Ethel Rogers
- Thomas Rogers

💡 **Indice 2 :** Hors ponctuation et mots vides, les phrases contenant le nom du coupable totalisent entre 500 et 1300 mots.

**Fonctions utiles :**

- Boucles `for` imbriquées
- Condition d'appartenance (`if` ... `in` ...)
- Traitement de texte (`nlp()`)
- Extraction de texte (`.text`)
- Mots vides (`.is_stop`)
- Ponctuation (`.is_punct`)
- Longueur (`len()`)

**Méthode :**

In [8]:
# Objectif : Compter le nombre de mots des phrases dans lesquelles les noms des suspects apparaissent

# Parcourir chaque suspect
    # Initialiser le compteur de mots des phrases dans lesquelles le nom du suspect apparaît
    # Parcourir chaque phrase
        # Si le nom du suspect est dans la phrase
            # Traiter la phrase comme un doc
            # Nettoyer la phrase (suppression des stopwords et de la ponctuation)
            # Ajouter la longueur de la phrase au nombre de mots

    # Affichage

**Solution :**

In [9]:
suspects = ["Armstrong", "Blore", "Brent", "Claythorne", "Lombard", "Macarthur", "Marston", "Wargrave"]

for suspect in suspects:
    nb_words_with_suspect = 0
    for sentence in sentences:
        if suspect in sentence:
            doc_sentence = nlp(sentence)
            cleaned_sentence = [token.text for token in doc_sentence if not token.is_stop and not token.is_punct]
            nb_words_with_suspect += len(cleaned_sentence)

    print(f"Nombre de mots contenant {suspect} : {nb_words_with_suspect}")

Nombre de mots contenant Armstrong : 1213
Nombre de mots contenant Blore : 1392
Nombre de mots contenant Brent : 888
Nombre de mots contenant Claythorne : 518
Nombre de mots contenant Lombard : 1582
Nombre de mots contenant Macarthur : 364
Nombre de mots contenant Marston : 412
Nombre de mots contenant Wargrave : 745


➡️ On peut éliminer :
- Blore
- Lombard
- Macarthur
- Marston

💡 **Indice 3 :** Hors ponctuation et mots vides, la partie du discours la plus fréquente dans les phrases contenant le nom du coupable n'est pas le verbe.

**Fonctions utiles :**

- Boucles `for` imbriquées
- Condition d'appartenance (`if` ... `in` ...)
- Traitement de texte (`nlp()`)
- Mots vides (`.is_stop`)
- Ponctuation (`.is_punct`)
- POS (`.pos_`)
- Dictionnaire (`{}`)
- Maximum (`max()`)

**Méthode :**

In [10]:
# Objectif : Compter les parties du discours des phrases dans lesquelles les noms des supects apparaissent et trouver la plus fréquente

# Parcourir chaque suspect
    # Parcourir chaque phrase
        # Si le nom du suspect est dans la phrase
            # Traiter la phrase comme un doc
            # Extraire les POS des tokens de la phrase (hors stopwords et ponctuation)
            # Initialiser un dictionnaire des comptes des POS
            # Parcourir chaque POS extraite
                # Si la POS appartient déjà au dictionnaire
                    # Ajouter 1 à sa fréquence
                # Si la POS n'appartient pas au dictionnaire
                    # Initialiser sa fréquence à 1
            # Extraire la POS qui a la fréquence la plus élevée

    # Affichage

**Solution :**

In [11]:
suspects = ["Armstrong", "Brent", "Claythorne", "Wargrave"]

for suspect in suspects:
    for sentence in sentences:
        if suspect in sentence:
            doc_sentence = nlp(sentence)
            pos_in_sent = [token.pos_ for token in doc_sentence if not token.is_stop and not token.is_punct]
            pos_count = {}
            for pos in pos_in_sent:
                if pos in pos_count:
                    pos_count[pos] += 1
                else:
                    pos_count[pos] = 1
            most_frequent_pos = max(pos_count, key=pos_count.get)

    print(f"Partie du discours la plus fréquente dans les phrases contenant {suspect} : {most_frequent_pos}")

Partie du discours la plus fréquente dans les phrases contenant Armstrong : NOUN
Partie du discours la plus fréquente dans les phrases contenant Brent : PROPN
Partie du discours la plus fréquente dans les phrases contenant Claythorne : VERB
Partie du discours la plus fréquente dans les phrases contenant Wargrave : NOUN


➡️ On peut éliminer :
- Claythorne

💡 **Indice 4 :** Hors ponctuation et mots vides, le lemme le plus fréquent dans les phrases contenant le nom du coupable ne commence pas par la lettre "_R_".

**Fonctions utiles :**

- Boucles `for` imbriquées
- Condition d'appartenance (`if` ... `in` ...)
- Traitement de texte (`nlp()`)
- Mots vides (`.is_stop`)
- Ponctuation (`.is_punct`)
- Lemmes (`.lemma_`)
- Dictionnaire (`{}`)
- Maximum (`max()`)

**Méthode :**

In [12]:
# Objectif : Compter les lemmes des phrases dans lesquelles les noms des supects apparaissent et trouver le plus fréquent

# Parcourir chaque suspect
    # Parcourir chaque phrase
        # Si le nom du suspect est dans la phrase
            # Traiter la phrase comme un doc
            # Extraire les lemmes des tokens de la phrase (hors stopwords et ponctuation)
            # Initialiser un dictionnaire des comptes des lemmes
            # Parcourir chaque lemme extrait
                # Si le lemme appartient déjà au dictionnaire
                    # Ajouter 1 à sa fréquence
                # Si le lemme n'appartient pas au dictionnaire
                    # Initialiser sa fréquence à 1
            # Extraire le lemme qui a la fréquence la plus élevée

    # Affichage

**Solution :**

In [13]:
suspects = ["Armstrong", "Brent", "Wargrave"]

for suspect in suspects:
    for sentence in sentences:
        if suspect in sentence:
            doc_sentence = nlp(sentence)
            lemmas = [token.lemma_ for token in doc_sentence if not token.is_stop and not token.is_punct]
            lemmas_count = {}
            for lemma in lemmas:
                if lemma in lemmas_count:
                    lemmas_count[lemma] += 1
                else:
                    lemmas_count[lemma] = 1
            most_frequent_lemma = max(lemmas_count, key=lemmas_count.get)

    print(f"Lemme le plus fréquent dans les phrases contenant {suspect} : {most_frequent_lemma}")

Lemme le plus fréquent dans les phrases contenant Armstrong : corps
Lemme le plus fréquent dans les phrases contenant Brent : Rogers
Lemme le plus fréquent dans les phrases contenant Wargrave : chambre


➡️ On peut éliminer :
- Brent

💡 **Indice 5 :** Le verbe dont le suspect est le plus souvent le sujet ne commence pas par la lettre "_D_".

**Fonctions utiles :**

- Boucles `for` imbriquées
- Condition d'appartenance (`if` ... `in` ...)
- Traitement de texte (`nlp()`)
- Relation de dépendance  (`.dep_`)
- Gouverneur syntaxique (`.head`)
- Lemmes (`.lemma_`)
- Dictionnaire (`{}`)
- Maximum (`max()`)

**Méthode :**

In [14]:
# Objectif : Trouver les mots auxquels les noms des suspects sont associés par la relation "nsubj", les compter et trouver le plus fréquent

# Parcourir chaque suspect
    # Initialiser un dictionnaire des comptes des verbes dont le suspect est le sujet
    # Parcourir chaque phrase
        # Si le nom du suspect est dans la phrase
            # Traiter la phrase comme un doc
            # Extraire le token de la phrase correspondant au nom du suspect
            # Si le nom du sujet a une relation de dépendance "nsubj" avec un verbe ("VERB")
                # Si le lemme de sa tête appartient déjà au dictionnaire
                    # Ajouter 1 à sa fréquence
                # Si le lemme de sa tête n'appartient pas au dictionnaire
                    # Initialiser sa fréquence à 1
    # Extraire le lemme qui a la fréquence la plus élevée

    # Affichage

**Solution :**

In [48]:
suspects = ["Armstrong", "Wargrave"]

for suspect in suspects:
    heads_count = {}
    for sentence in sentences:
        if suspect in sentence:
            doc_sentence = nlp(sentence)
            ref_token = [token for token in doc_sentence if token.text == suspect][0]
            if ref_token.dep_ == "nsubj" and ref_token.head.pos_ == "VERB":
                if ref_token.head.lemma_ in heads_count:
                    heads_count[ref_token.head.lemma_] += 1
                else:
                    heads_count[ref_token.head.lemma_] = 1
    most_frequent_head = max(heads_count, key=heads_count.get)

    print(f"Verbe dont {suspect} est le plus souvent le sujet : {most_frequent_head}")

Verbe dont Armstrong est le plus souvent le sujet : dire
Verbe dont Wargrave est le plus souvent le sujet : hocher


➡️ On peut éliminer :
- Armstrong

**Le coupable est donc...**

<img src="https://st2.depositphotos.com/2101611/6966/v/450/depositphotos_69669865-stock-illustration-mugshot-of-anonymous-male-silhouette.jpg" alt="Stopwords" width="300">

**Le juge Wargrave !**

<img src="https://thumbs.dreamstime.com/b/juge-17648664.jpg?w=768" alt="Stopwords" width="200">

- Nombre de phrases dans lesquelles son nom apparaît : 113
- Nombre de mots hors mots-vides et ponctuations : 745
- Partie du discours la plus fréquente : NOUN
- Lemme le plus fréquent : chambre
- Verbe dont le suspect est le plus souvent le sujet : prendre